File: algorithms/algorithms-cplusplus/leetcode/Question_1080_Insufficient_Nodes_in_Root_to_Leaf_Paths.cc

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
 public:
  TreeNode* sufficientSubset(TreeNode* root, int limit) {
    bool valid = help(root, limit, 0);
    return valid ? root : nullptr;
  }

  bool help(TreeNode* root, int limit, int pre) {
    if (!root) {
      return false;
    }
    pre += root->val;
    if (root->left == nullptr && root->right == nullptr) {
      return pre >= limit;
    }
    bool left = help(root->left, limit, pre);
    bool right = help(root->right, limit, pre);
    if (!left) {
      root->left = nullptr;
    }
    if (!right) {
      root->right = nullptr;
    }
    if (left || right) {
      return true;
    }
    return false;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_1080_Insufficient_Nodes_in_Root_to_Leaf_Paths.java
package com.lanceolata.coding.algorithms.java.leetcode;

public class Question_1080_Insufficient_Nodes_in_Root_to_Leaf_Paths {
    // Definition for a binary tree node.
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public TreeNode sufficientSubset(TreeNode root, int limit) {
        return help(root, limit, 0) ? root : null;
    }

    private boolean help(TreeNode root, int limit, int pre) {
        if (root == null) {
            return false;
        }
        pre += root.val;
        if (root.left == null && root.right == null) {
            return pre >= limit;
        }
        boolean left = help(root.left, limit, pre);
        boolean right = help(root.right, limit, pre);
        if (!left) {
            root.left = null;
        }
        if (!right) {
            root.right = null;
        }
        if (!left && !right) {
            return false;
        }
        return true;
    }
}
